woocommerce/woocommerce
# Add File Headers filter at 'plugins_loaded' hook so that WooCommerce Helper can detect all locally installed Woo plugins

Fixes #21668

**Description of the bug:**

If any plugin calls the get_plugins function before the woocommerce_init action gets called, the WooCommerce Helper doesn't detect some locally installed Woo plugins. This leads to customers not being able to activate WooCommerce.com subscriptions for those plugins.

**Steps to Reproduce:**

I've been able to reproduce this behaviour on my local environment.

- I installed and activated the WooCommerce Multi Currency plugin on my local environment.

- I created a dummy plugin with the following code and activated it:

  ```
  <?php
  
  /*
  Plugin Name: Kanishk Test
  Plugin URI: http://my-awesomeness-emporium.com
  description: a plugin to create awesomeness and spread joy
  Version: 1.2
  Author: Mr. Awesome
  Author URI: http://mrtotallyawesome.com
  License: GPL2
  */

  include_once( ABSPATH . 'wp-admin/includes/plugin.php' );
  add_filter( 'before_woocommerce_init', 'get_plugins' );
  ```

- Even though I have WooCommerce.com subscriptions for both `WooCommerce Multi Currency` and `WooCommerce bookings`, I can only activate the `WooCommerce Bookings` subscription.

  ![screen shot on 2018-10-24 at 23-26-22](https://user-images.githubusercontent.com/3773215/47492989-db786180-d86b-11e8-9604-5aa60f3bcd37.png)

- If I de-activate my dummy plugin, then everything works fine.

**Cause:**

I've been able to debug this on my local environment and it looks like this is the cause of the behaviour:

- The `WooCommerce Helper` adds a filter on the `extra_plugins_header` filter so that the `Woo: ProductID:DownloadID` header can be parsed from the plugin's file. This filter is added when the Helper is loaded and the Helper is loaded on the `init` hook.

- If another plugin calls `get_plugins()` before the above filter gets hooked, `get_plugins` doesn't get to know about the filter and returns a list of plugins with their headers (without the Woo header). This value gets cached by get_plugins.

- Now, later on in the lifecycle, the filter gets added *but* subsequent calls to `get_plugins()` still don't return the `Woo` header since `get_plugins` returns the cached information.

- Because the `Woo` header isn't fetched while fetching a list of plugins, the Helper cannot detect it as a locally installed Woo plugin and therefore, doesn't show an option to activate the subscription.

- This is only occurring for `WooCommerce Multi Currency` plugin since it uses the file headers to pass the Woo header information to the helper. Other plugins like WooCommerce Bookings use our legacy mechanism (`woothemes_queue_update`) to supply the header to us.

**Proposed fix:**

- Added a new class named `WC_Helper_File_Headers` which is responsible for adding the `extra_headers` filter on the `extra_plugin_headers` and `extra_theme_headers` hooks.

- Removed this functionality from the main `WC_Helper` class

- Load the `WC_Helper_File_Headers` class on `plugins_loaded` rather than at `init`.

**Testing instructions:**

- Purchase the `WooCommerce Multi Currency` plugin on `WooCommerce.com`

- Connect your local WooCommerce development environment Helper to the WooCommerce.com

- Install and activate the `WooCommerce Multi Currency` and `WooCommerce Bookings` plugin on your local environment.

- Create a dummy plugin with the following code and activate it:

  ```
  <?php
  
  /*
  Plugin Name: Kanishk Test
  Plugin URI: http://my-awesomeness-emporium.com
  description: a plugin to create awesomeness and spread joy
  Version: 1.2
  Author: Mr. Awesome
  Author URI: http://mrtotallyawesome.com
  License: GPL2
  */

  include_once( ABSPATH . 'wp-admin/includes/plugin.php' );
  add_filter( 'before_woocommerce_init', 'get_plugins' );
  ```

- You should be able to activate WooCommerce.com subscriptions for both plugins

  ![screen shot on 2018-10-24 at 23-24-13](https://user-images.githubusercontent.com/3773215/47493578-3b233c80-d86d-11e8-9b15-2f1995a8ddcd.png)

cc @mattyza @DanReyLop
